Initial Version

hefroy opened the pull request.

Commit:
Import files from meta-scooteros project

Commit:
Add pwm_phase_offset param

Commit:
Add adaptive mode

Commit:
Use the SDMA r2 register to store the current duty

Commit:
Simplify SDMA script

Commit:
Adjust binary search alg for leftmost value

Commit:
Correct debug info

Commit:
Add active functionality

Commit:
Load SDMA script on channel 0 only

Also tidy up a bit

Commit:
Move all shared init/deinit to channel 0

Commit:
Tidy up

Commit:
Add more locking

Commit:
Improve doc

Commit:
Remove last FIXMEs, write README

gmphillips commented on .editorconfig:
```diff
@@ -0,0 +1,8 @@
+root = true
```
Test finding

hefroy commented on linux-imx-4.9.88/git/arch/arm/mach-imx/epit_api.c:
```diff
@@ -0,0 +1,333 @@
+/**
```
It would be better to either:
- Just provide .patch files
- or commit the original files, then perform another commit with the changes.

hefroy replied:
Fixed by only providing .patch files

drubin commented on README.md:
```diff
@@ -1 +1,232 @@
-# kernel-module-imx-pwm-led
+# Freescale i.MX PWM LED Linux Kernel Module
```
Is this standard? normally copyright and license are always at the end below Overview?

hefroy replied:
No it's not standard, agreed I'll move it

hefroy replied:
Fixed by removing the license info from the README, it's already in multiple other places and was cluttering the README

drubin replied:
Epic. sorry I can't really review any of the other stuff.  :)

Commit:
Fix several issues, improve doc

Issues were:
- Active/inactive were inverted
- ioctl_set_duty did not start the LED
- Prevent fade player reference count being left as non-zero, either by:
  - Cue referencing the same LED more than once
  - Failure during cue load, leaving loaded fades that are not started
- Prevent deadlocks due to order of LED locking
Also:
- Add Kconfig and Makefile patches

Commit:
Remove license info from README

since it's already mentioned in multiple other places.
Also:
- Move the quick-start up to the top
- Add a bit more info in the overview about fades and cues
- Rename cue items to cue actions since that's what they are

Commit:
Fix find_fade_dir

Commit:
Add perm_mode param

Commit:
Add support for setting pwm repeat

Commit:
Add preload workaround for hardware bug

Commit:
Improve documentation and warnings

hefroy commented on README.md:
```diff
@@ -1 +1,247 @@
+
+do_install_append() {
+    install -d ${D}${includedir}/linux
+    install -m 0644 ${S}/pwm_led.h ${D}${includedir}/linux
```
Since this isn't a portable module, the include directory should probably be `<imx/linux/pwm_led.h>` rather than `<linux/pwm.h>`

hefroy replied:
Fixed

hefroy commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+	}
+	self->pwm_fifo_period_us =
+		(period * (prescaler + 1) << repeat) / (clk / 1000000);
+	dev_dbg(self->dev, "%s: setting PWMPR: 0x%08x (%u.%03u Hz)", __func__,
```
The `repeat` value should also be logged.

hefroy commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+	uint invert = (arg & PWM_LED_CFG_MASK_INVERT) >> PWM_LED_CFG_BIT_INVERT;
+	uint repeat = (arg & PWM_LED_CFG_MASK_REPEAT) >> PWM_LED_CFG_BIT_REPEAT;
+	dev_dbg(self->dev,
+		"ioctl: CONFIGURE: period: %u, prescaler: %u, invert: %u",
```
The `repeat` value should also be logged.

hefroy replied:
Fixed

hefroy commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+			return 0;
+		}
+		/* Sleep until the FIFO should be empty: */
+		min_sleep = level * self->pwm_fifo_period_us;
```
Should probably have an upper bound for the sleep, you wouldn't want to sleep for more than 100ms.

hefroy replied:
Fixed

odurc commented on README.md:
```diff
@@ -1 +1,247 @@
+  IPG clock is 24 MHz and the PWM period is 12000, the output PWM frequency
+  will be 2 kHz.
+- A PWM 'duty-cycle value' is defined as a 16-bit counter value at which the
+  output changes from 1 to 0, which can range from zero to the PWM period value.
```
> the output changes from 1 to 0

Does that mean the outputs are normally on by default?

hefroy replied:
I mean the output changes from 0 to 1 when the counter wraps round from pwm_period - 1 to zero, and 1 to 0 when it passes the duty cycle value. Do you think it's not clear enough? I can add this to make it clearer.

odurc replied:
Got it now. I guess it's clear enough.

hefroy commented on imx_pwm_led_sdma.asm:
```diff
@@ -0,0 +1,78 @@
+    cmpeqi r6, 0 ; if ($preload == 0)         // If the preload flag is not set:
+    bt load      ;     goto load;             // Jump to load
+    ldi r6, 0    ; $preload = 0;              // Clear the preload flag
+    cmpeqi r4, 0 ; if ($active == 0)          // If inactive:
+    bt preload0  ;     goto preload0;         // Jump to preload0
+    stf r2, 0x2b ; PWMSAR = $duty             // Set the current PWM duty
+    btsti r1, 0  ; if ($one & 0x01)           // If true: No relative jmp exists
+    bt load      ;     goto load;             // Jump to load
+preload0:
+    stf r4, 0x2b ; PWMSAR = $active;          // Set the current PWM duty (zero)
```
It's probably better just to read the current duty cycle from the PWMSAR, as you could have just changed the $active flag, meaning the first value you load is not the current duty cycle.

hefroy replied:
Actually this isn't possible since the LED is locked when the active status is being changed.
Using the r2 value is faster than reading the value from PWMSAR too.

hefroy commented on README.md:
```diff
@@ -1 +1,247 @@
+- unu,sdma-script-origin : **Only required for the first device**
+  Location to load the SDMA script (in 32-bit data words).
+  The 4KB SDMA internal RAM area for scripts and data is located at
+  0xc00 - 0xfff. The imx-pwm-led script is 10 words long.
```
The script is longer now

hefroy replied:
Fixed

odurc commented on README.md:
```diff
@@ -1 +1,247 @@
+      the fade will be started at the point closest to 50%, rather than first
+      jumping to 100%.
+    - 'Active': This can be changed at run time. When a channel is active, the
+      values of a fade played on the channel will set be set as the channel's
```
typo: "will set be set"

hefroy replied:
Fixed

odurc commented on pwm_led.h:
```diff
@@ -0,0 +1,67 @@
+#define PWM_LED_CUE_ACTION_MASK_TYPE 0x00000F00 /* See PWM_LED_CUE_ACTION_TYPE_x */
+#define PWM_LED_CUE_ACTION_MASK_VAL  0xFFFF0000
+#define PWM_LED_CUE_ACTION_TYPE_FADE 0 /* Value=fade index */
+#define PWM_LED_CUE_ACTION_TYPE_DUTY 1 /* Value=duty cycle */
```
maybe those types could be an enum?

hefroy replied:
Fixed

odurc commented on README.md:
```diff
@@ -1 +1,247 @@
+    PWM_LED_OPEN_CUE   - opens the cue of given index
+    PWM_LED_CLOSE_FADE - closes the fade of given index
+    PWM_LED_CLOSE_CUE  - closes a cue of given index
+    PWM_LED_PLAY_FADE  - plays a fade of given index
```
what is the difference between playing a fade using this command or using a cue?

>A 'cue' is a list of actions to start on one or more channels at the same time. The supported actions are:
    Starting playback of a fade on a given channel.
    Setting a given duty-cycle directly on a given channel.

hefroy replied:
Using this command will play the fade on just the channel that is targeted with the `ioctl`.
Using a cue gives you the option to play the fade on multiple channels at the same time.

odurc commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+};
+
+/** Module global data struct */
+struct glb_data {
```
this type holds the data for all the devices, right? and you are implementing some sort of device (instances) manager. I'm just wondering if it worth to rename it to something more descriptive on that line.

cosmetic: any reason to prefer glb_ over g_?

hefroy replied:
Agreed, I'll rename it as `static struct module_data g_data;`. `module_data` indicates that data is used by all devices, and `g_data` is nice and short, since this is used in a lot of places.

hefroy replied:
Fixed

odurc commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+static u32 max_leds         = 8; /* imx6ul has 8 PWMs */
+static u32 max_fade_size    = 4096; /* DMA memory block size is 4KB */
+static u32 sdma_priority    = 5; /* Allowed range: 1 to 6 */
+static u32 perm_mode        = PERM_MODE_PID_OWNER; /* Permission mode */
```
I assume you are not prefixing those variables with g_ because these are the same names used as the module parameters?

another cosmetic comment. this one specifically (`perm_mode`) got me confused for a 1 second because there is a enum also named as `perm_mode`.

hefroy replied:
Yes exactly the `module_param` macro makes the variable visible as a module parameter with exactly that name.
Yes it would be nice to have typedef and a `_t` suffix, but the linux coding guideline says not to use typedef, rather enum/struct before the tag name.

odurc commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+		ret = -EPERM;
+		goto out;
+	}
+	fade->owner_pid = 0;
```
I don't think that is a real problem, but the PID 0 is the kernel scheduler. Maybe you can set it -1, but that would break some of your *if* tests.

hefroy replied:
Yes I also figured this won't be a problem, as something is seriously wrong if the the scheduler is making calls to this module.

odurc commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+	/* Enable clocks */
+	int ret = 0;
+
+	if ((ret = clk_prepare_enable(self->clk_per))) {
```
suggestion: keep variable assignment operation outside of test statements for improving readability.

I found quite a few with `grep -Irn "if (.* = .*"`

hefroy replied:
I was going along with the kernel coding style which seems to encourage this, agreed that it's not normally good practice.
(GCC gives a warning if the extra brackets are missing.)

odurc commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+		dev_warn(self->dev, "%s: fade %u: len %u < %u", __func__,
+			 fade_idx, fade->len, sizeof(u16));
+		ret = -ENODATA;
+		goto out;
```
does the fade need to be closed in case it fails here?

hefroy replied:
No it should be left open for several reasons:
- if `PERM_MODE_PID_OWNER` is used and there are multiple processes handling different LED devices, then it's up to the system designer to assign which fades/cues are being used by which process. Hence there's no need to close the fade/cue, since nobody else should be using it.
- calling open_fade/cue more than once is ok if the process owns the fade/cue, it's really just assigning the calling process as the owner (if `PERM_MODE_PID_OWNER` is used).
- when the device is closed all fades and cues that were opened by the calling process are automatically closed.

odurc commented on imx_pwm_led.c:
```diff
@@ -0,0 +1,2069 @@
+		dev_warn(self->dev, "%s: fade %u: len %u %% %u != 0", __func__,
+			 fade_idx, fade->len, sizeof(u16));
+		ret = -EFAULT;
+		goto out;
```
does the fade need to be closed in case it fails here?

hefroy replied:
Same as above

Commit:
Fix review findings

Commit:
Change to .patch files for kernel patches

Instead of including the whole modified source files.
